lib/kubernetes-deploy/kubernetes_resource.rb

```diff
@@ -54,19 +42,27 @@ def timeout
       self.class.timeout
     end
 
-    def initialize(name:, namespace:, context:, file:, logger:)
+    def initialize(namespace:, context:, definition:, logger:)
       # subclasses must also set these if they define their own initializer
-      @name = name
+      unless @name = definition.fetch("metadata", {})["name"]
```

Comment:
It looks like this doesn't handle the case when `name` is an empty string? Could probably use `present?` from ActiveSupport